autoware_reference_system/test/generate_traces.py

```diff
@@ -73,7 +73,24 @@ def generate_test_description():
         ] + DEFAULT_CONTEXT,
     )
 
-    launch_description.add_action(trace_action)
+    trace_memalloc_action = Trace(
+        session_name='profile_' + test_exe_name + '_' + str(timeout) + 's',
+        events_ust=[
+            'lttng_ust_libc:malloc',
+            'lttng_ust_libc:calloc',
+            'lttng_ust_libc:realloc',
+            'lttng_ust_libc:free',
+            'lttng_ust_libc:memalign',
+            'lttng_ust_libc:posix_memalign'
+        ] + DEFAULT_EVENTS_ROS,
+        events_kernel=[
+            'kmem_mm_page_alloc',
+            'kmem_mm_page_free'
+        ]
+    )
+
+    # launch_description.add_action(trace_callback_action)
+    launch_description.add_action(trace_memalloc_action)
```

Comment:
@christophebedard would love your input here. when trying to generate a report for memory usage I keep getting a `missing events` error when processing the generated traces. I tried adding [the same events you have here] in the memory usage example but I am still seeing the error. any advice?

Reply:
specifically these are the missing events:

```
missing events: {'UserspaceMemoryUsageHandler': {'lttng_ust_libc:malloc', 'lttng_ust_libc:free'}}
```

Reply:
also @christophebedard  are you allowed to add multiple `Trace` actions to one single `LaunchDescription`?

Reply:
> @christophebedard would love your input here. when trying to generate a report for memory usage I keep getting a `missing events` error when processing the generated traces. I tried adding [the same events you have here](https://gitlab.com/ros-tracing/tracetools_analysis/-/blob/master/tracetools_analysis/launch/memory_usage.launch.py) in the memory usage example but I am still seeing the error. any advice?

make sure you have `liblttng-ust-libc-wrapper.so`. It's under `/usr/lib/x86_64-linux-gnu/liblttng-ust-libc-wrapper.so` on my system. The `Trace` action should find the full path using `whereis -b liblttng-ust-libc-wrapper.so` and then make sure it's `LD_PRELOAD`ed, but something is wrong if you have no `lttng_ust_libc:*` events. Maybe enable `--debug` for the launch files and see if there's a warning (like "Could not find shared library").

Reply:
> also @christophebedard are you allowed to add multiple `Trace` actions to one single `LaunchDescription`?

it should work yes, but it's probably unnecessary here.

Reply:
> make sure you have liblttng-ust-libc-wrapper.so
confirmed i've got it and `whereis` can find it just fine. ill try the `--debug` setting now

Reply:
I've merged the two `Trace` actions into a single action, and running `ros2 test build/autoware_reference_system/generate_traces_autoware_default_singlethreaded_rmw_cyclonedds_cpp_5s.py --verbose` shows that the two shared libraries are found:

```
[DEBUG] [tracetools_launch.actions.ld_preload]: Shared library found at: /usr/lib/x86_64-linux-gnu/liblttng-ust-cyg-profile-fast.so
[DEBUG] [tracetools_launch.actions.ld_preload]: Shared library found at: /usr/lib/x86_64-linux-gnu/liblttng-ust-libc-wrapper.so
```

For the profiling tracepoints (`*:func_{entry,exit}`), you need to build with `
`-finstrument-functions``: https://gitlab.com/ros-tracing/tracetools_analysis/-/blob/3fb2c697e48e9bd3c8421c31eec02e84995947e6/tracetools_analysis/tracetools_analysis/processor/profile.py#L40

Reply:
For the `lttng_ust_libc:{malloc,free}` events, I'm not sure why they're not showing up. Can you try running this launch file: https://gitlab.com/ros-tracing/tracetools_analysis/-/blob/master/tracetools_analysis/launch/memory_usage.launch.py. And then running this notebook: https://gitlab.com/ros-tracing/tracetools_analysis/-/blob/master/tracetools_analysis/analysis/memory_usage.ipynb. Those work for me. Is there any way you can switch the trace generation script/test so that it's not using `add_ros_test`?

Reply:
argh after doing a bit of trial and error, it seems to be due to the `env=env` option in `ExecuteProcess`. If I comment it out in `generate_traces.py`, re-build, and re-run the tests to re-generate the traces, I do finally get `*:func_{entry,exit}` events.

I'm guessing the [`LdPreload` action](https://gitlab.com/ros-tracing/ros2_tracing/-/blob/master/tracetools_launch/tracetools_launch/actions/ld_preload.py) doesn't mix well with using the `env=` option explicitly. Maybe try to use `launch.actions.SetEnvironmentVariable` actions (added to the launch description first?) to set `RCUTILS_CONSOLE_OUTPUT_FORMAT`, `RCL_ASSERT_RMW_ID_MATCHES`, and `RMW_IMPLEMENTATION`?

Reply:
@christophebedard awesome!!! thanks for digging into this. I'll merge the traces together and try your advice on the `env` topics now.

p.s. also cool to hear that you were able to use this branch to generate some traces 😄

Reply:
@christophebedard yeah that was it - im not seeing the `missing events` error anymore....now I can focus on the report generation.

let me know if there are any specific KPIs you think we should focus on, and thanks again for the support here.

Reply:
as we discussed last week, we can start with frequency/jitter (say, mean rate and standard deviation?) for subscription callbacks (wrt the start of callbacks) and publications (wrt `ros2:rclcpp_publish`).

I don't think `tracetools_analysis` currently fully supports the `ros2:*_publish` tracepoints even in Rolling, so I'll need to take care of that first. But the callback data is all there.